TESTS/mbedmicro-rtos-mbed/signals/main.cpp

```diff
@@ -15,49 +15,357 @@
+    int32_t ret;
+    Thread tb(osPriorityNormal, TEST_STACK_SIZE);
+    tb.start(callback(run_signal_wait2<ALL_SIGNALS, osWaitForever>));
+
+    sem_a.wait();
+    TEST_ASSERT_EQUAL(Thread::WaitingThreadFlag, tb.get_state());
+
+    ret = tb.signal_set(ALL_SIGNALS);
+    TEST_ASSERT_EQUAL(0, ret);
+
+    sem_a.wait();
+    TEST_ASSERT_EQUAL(osEventSignal, ret_status);
+}
 
-    while (true) {
-        Thread::wait(2 * SIGNAL_HANDLE_DELEY);
-        thread.signal_set(SIGNAL_SET_VALUE);
-        if (signal_counter == SIGNALS_TO_EMIT) {
-            printf("Handled %d signals\r\n", signal_counter);
-            result = true;
-            break;
-        }
+/** Test signal_wait on thread with no signals already set - signals set one by one in loop
+
+    Given two threads A and B running and B waiting for a thread signals to be set
+    When thread A executes @a signal_set(signal) in loop with all possible signals
+    Then @a signal_set status should be current set signals or 0 if last remaining signal set
+         thread B @a signal_wait status is osEventSignal indicating a all wait signals was set
+ */
+void test_wait_all_loop(void)
+{
+    int32_t ret;
+    Thread tb(osPriorityNormal, TEST_STACK_SIZE);
+    tb.start(callback(run_signal_wait2<ALL_SIGNALS, osWaitForever>));
+
+    sem_a.wait();
+    int32_t signals = 0;
+    for (int i = 0; i < MAX_FLAG_POS; i++) {
+        int32_t signal = 1 << i;
+        ret = tb.signal_set(signal);
+        signals |= signal;
+        TEST_ASSERT_EQUAL(signals, ret);
     }
-    GREENTEA_TESTSUITE_RESULT(result);
-    return 0;
+    ret = tb.signal_set(1 << MAX_FLAG_POS);
+    TEST_ASSERT_EQUAL(0, ret);
+    sem_a.wait();
+    TEST_ASSERT_EQUAL(osEventSignal, ret_status);
+}
+
+
+utest::v1::status_t test_setup(const size_t number_of_cases)
+{
+    GREENTEA_SETUP(10, "default_auto");
+    return verbose_test_setup_handler(number_of_cases);
+}
+
+Case cases[] = {
+    Case("Test signal_clr with no signals specified", test_clear_0),
+    Case("Test if any signals are set on just created thread", test_init_state),
+    Case("Test signal_set with all signals set in one shot", test_set_all),
+    Case("Test signal_set with prohibited signal", test_set_prohibited),
+    Case("Test signal_clr with all signals clear in one shot", test_clear_all),
+    Case("Test signal_set with all signals set one after the other in loop", test_set_all_loop),
+    Case("Test signal_wait with no signals specified and 0 timeout", test_wait_0),
+    Case("Test signal_wait with 1[ms] timeout and no signal set", test_wait_timeout<0, 1>),
+    Case("Test signal_wait with 1[ms] timeout and all signals set", test_wait_timeout<ALL_SIGNALS, 1>),
+    Case("Test signal_wait on thread with all signals already set", test_wait_all_already_set),
+    Case("Test signal_wait on thread with no signals already set", test_wait_all),
+    Case("Test signal_wait on thread with no signals already set - signals set one by one in loop", test_wait_all_loop)
```

Comment:
Maybe we could also validate that:
* we can set a signal then clear or wait and then set it again
* what happens when we set the same signal twice